presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractMinMaxBy.java

```diff
@@ -96,10 +100,27 @@ public InternalAggregationFunction specialize(BoundVariables boundVariables, int
     private InternalAggregationFunction generateAggregation(Type valueType, Type keyType, FunctionRegistry functionRegistry)
     {
         Class<?> stateClazz = getStateClass(keyType.getJavaType(), valueType.getJavaType());
-        Map<String, Type> stateFieldTypes = ImmutableMap.of("First", keyType, "Second", valueType);
         DynamicClassLoader classLoader = new DynamicClassLoader(getClass().getClassLoader());
-        AccumulatorStateFactory<?> stateFactory = StateCompiler.generateStateFactory(stateClazz, stateFieldTypes, classLoader);
-        AccumulatorStateSerializer<?> stateSerializer = StateCompiler.generateStateSerializer(stateClazz, stateFieldTypes, classLoader);
+
+        // Generate states and serializers:
+        // For value that is a Block or Slice, we store them as Block/position combination
+        // to avoid generating long-living objects through getSlice or getObject.
+        // On the other hand, keys can have the same problem. But usually they are primitive types (given the nature of comparison).
+        AccumulatorStateFactory<?> stateFactory;
+        AccumulatorStateSerializer<?> stateSerializer;
+        if (isJavaTypePrimitive(valueType)) {
+            Map<String, Type> stateFieldTypes = ImmutableMap.of("First", keyType, "Second", valueType);
+            stateFactory = StateCompiler.generateStateFactory(stateClazz, stateFieldTypes, classLoader);
+            stateSerializer = StateCompiler.generateStateSerializer(stateClazz, stateFieldTypes, classLoader);
+        }
+        else {
+            // StateCompiler checks type compatibility; need to make sure SecondBlock is a Block rather than a Slice (defined by valueType) in some cases.
+            // Given "Second" in this case is always a Block, we only need to make sure the getter and setter of the Blocks are properly generated.
+            // No getter or setter for Slice is needed for AccumulatorState.
+            stateFactory = StateCompiler.generateStateFactory(stateClazz, ImmutableMap.of("First", keyType, "SecondBlock", new ArrayType(valueType)), classLoader);
```

Comment:
Use `new ArrayType(valueType))` looks a bit hacky.

An alternative way would be making `generateStateFactory` also takes a list of boolean, indicate whether the ith field is represented by Block+Position.

Any thoughts? @haozhun

Reply:
- Add comments to explain we are generating state class but not serializer.
- Add comments why we want to use `ArrayType(valueType)`.